src/par_iter/find_first_last.rs

```diff
@@ -0,0 +1,319 @@
+    where ITEM: Send,
+          FIND_OP: Fn(&ITEM) -> bool + Sync
+{
+    fn split_off(&self) -> Self {
+        // Upper bound for one consumer will be lower bound for the other. This
+        // overlap is okay, because only one of the bounds will be used for
+        // comparing against best_found; the other is kept only to be able to
+        // divide the range in half.
+        //
+        // When the resolution of usize has been exhausted (i.e. when
+        // upper_bound = lower_bound), both results of this split will have the
+        // same range. When that happens, we lose the ability to tell one
+        // consumer to stop working when the other finds a better match, but the
+        // reducer ensures that the best answer is still returned (see the test
+        // above).
+        //
+        // This code assumes that the caller of split_off will use the result as
+        // the *left* side of this iterator, and the remainder of self as the
+        // *right* side.
+        let old_lower_bound = self.lower_bound.get();
+        let median = old_lower_bound + ((self.upper_bound - old_lower_bound) / 2);
+        self.lower_bound.set(median);
+
+        FindConsumer {
+            find_op: self.find_op,
+            lower_bound: Cell::new(old_lower_bound),
+            upper_bound: median,
+            match_position: self.match_position,
+            best_found: self.best_found,
+        }
+    }
+
+    fn to_reducer(&self) -> Self::Reducer {
+        FindReducer { match_position: self.match_position }
+    }
+}
+
+struct FindFolder<'f, ITEM, FIND_OP: 'f> {
+    find_op: &'f FIND_OP,
+    boundary: usize,
+    match_position: MatchPosition,
+    best_found: &'f AtomicUsize,
+    item: Option<ITEM>,
+}
+
+impl<'f, FIND_OP: 'f + Fn(&ITEM) -> bool, ITEM> Folder<ITEM> for FindFolder<'f, ITEM, FIND_OP> {
+    type Result = Option<ITEM>;
+
+    fn consume(mut self, item: ITEM) -> Self {
+        let found_best_in_range = match self.match_position {
+            MatchPosition::Leftmost => self.item.is_some(),
+            MatchPosition::Rightmost => false,
+        };
+
+        if !found_best_in_range && (self.find_op)(&item) {
+            // Continuously try to set best_found until we succeed or we
+            // discover a better match was already found.
+            let mut current = self.best_found.load(Ordering::Relaxed);
+            let mut exchange_result = Result::Err(0);
+            while !better_position(current, self.boundary, self.match_position) &&
+                  exchange_result.is_err() {
+                exchange_result = self.best_found.compare_exchange_weak(current,
+                                                                        self.boundary,
+                                                                        Ordering::Relaxed,
+                                                                        Ordering::Relaxed);
+                if exchange_result.is_err() {
```

Comment:
Actually, `compare_exchange_weak` gives you back the value that you failed to swap in its `Err` variant. So you could rewrite this as:

```rust
let mut current = self.best_found.load(Ordering::Relaxed);
loop {
    if better_position(current, self.boundary, self.match_position) {
        break;
    }
    match self.best_found.compare_exchange_weak(current, self.boundary, Relaxed, Relaxed) {
        Ok(_) => {
            self.item = Some(item);
            break;
        }
        Err(v) => current = v,
    }
}
```

This seems mildly clearer (and more efficient) to me (I also prefer the `loop+break` formulation over the `while` loop; it's a nit, but I find it easier for me to parse what is happening).